x/crosschain/keeper/msg_server_whitelist_erc20.go

```diff
@@ -0,0 +1,122 @@
+	if err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrDeployContract, "failed to deploy ZRC20 contract for ERC20 contract address (%s) on chain (%d)", msg.Erc20Address, msg.ChainId)
+	}
+	if zrc20Addr == (ethcommon.Address{}) {
+		return nil, sdkerrors.Wrapf(types.ErrDeployContract, "deployed ZRC20 return 0 address for ERC20 contract address (%s) on chain (%d)", msg.Erc20Address, msg.ChainId)
+	}
+	// add to the foreign coins
+	foreignCoin := fungibletypes.ForeignCoins{
+		Zrc20ContractAddress: zrc20Addr.Hex(),
+		Asset:                msg.Erc20Address,
+		ForeignChainId:       msg.ChainId,
+		Decimals:             msg.Decimals,
+		Name:                 msg.Name,
+		Symbol:               msg.Symbol,
+		CoinType:             common.CoinType_ERC20,
+		GasLimit:             uint64(msg.GasLimit),
+	}
+
+	param, found := k.zetaObserverKeeper.GetCoreParamsByChainID(ctx, msg.ChainId)
+	if !found {
+		return nil, sdkerrors.Wrapf(types.ErrInvalidChainID, "core params not found for chain id (%d)", msg.ChainId)
+	}
+
+	medianGasPrice, isFound := k.GetMedianGasPriceInUint(ctx, msg.ChainId)
+	if !isFound {
+		return nil, sdkerrors.Wrapf(types.ErrUnableToGetGasPrice, "median gas price not found for chain id (%d)", msg.ChainId)
+	}
+	medianGasPrice = medianGasPrice.MulUint64(2) // overpays gas price by 2x
+
+	hash := tmbytes.HexBytes(tmtypes.Tx(ctx.TxBytes()).Hash())
+
+	index := crypto.Keccak256Hash(hash.Bytes())
+	cctx := types.CrossChainTx{
+		Creator:        msg.Creator,
+		Index:          index.String(),
+		ZetaFees:       sdk.NewUint(0),
+		RelayedMessage: fmt.Sprintf("%s:%s", common.CmdWhitelistERC20, msg.Erc20Address),
+		CctxStatus: &types.Status{
+			Status:              types.CctxStatus_PendingOutbound,
+			StatusMessage:       "",
+			LastUpdateTimestamp: 0,
+		},
+		InboundTxParams: &types.InboundTxParams{
+			Sender:                          "",
+			SenderChainId:                   0,
+			TxOrigin:                        "",
+			CoinType:                        common.CoinType_Cmd,
+			Asset:                           "",
+			Amount:                          sdk.Uint{},
+			InboundTxObservedHash:           hash.String(), // all Upper case Cosmos TX HEX, with no 0x prefix
+			InboundTxObservedExternalHeight: 0,
+			InboundTxBallotIndex:            "",
+			InboundTxFinalizedZetaHeight:    0,
+		},
+		OutboundTxParams: []*types.OutboundTxParams{
+			&types.OutboundTxParams{
+				Receiver:                         param.Erc20CustodyContractAddress,
+				ReceiverChainId:                  msg.ChainId,
+				CoinType:                         common.CoinType_Cmd,
+				Amount:                           sdk.NewUint(0),
+				OutboundTxTssNonce:               0,
+				OutboundTxGasLimit:               100_000,
+				OutboundTxGasPrice:               medianGasPrice.String(),
+				OutboundTxHash:                   "",
+				OutboundTxBallotIndex:            "",
+				OutboundTxObservedExternalHeight: 0,
+			},
+		},
+	}
```

Comment:
Seems like most fields are initialized with default or null value.
I think we would benefit from implementing a `types.NewCrossChainTx(...)` initializer.

Reply:
This is intentional. I actually like it this way because in a single place, you know how the CCTX struct will look like.

Reply:
We do have somekind of NewCrossChainTx() but i don't like to use it here. It's not flexible enough and yet hides the filled fields inside another function.

Reply:
Ok, I find it a bit extensive IMO but that sounds fair